File: Arrays/414. Third Maximum Number/414. Third Maximum Number.py
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        first_max = second_max = third_max = -float("inf")

        for n in nums:
            if n in (first_max, second_max, third_max):
                continue

            if n > first_max:
                first_max, second_max, third_max = n, first_max, second_max
            elif n > second_max:
                second_max, third_max = n, second_max
            elif n > third_max:
                third_max = n
        
        return third_max if third_max != -float("inf") else first_max

# Time: O(n)
# Space: O(1)

File: Arrays/414. Third Maximum Number/414. Third Maximum Number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var thirdMax = function (nums) {
  let first_max = -Infinity;
  let second_max = -Infinity;
  let third_max = -Infinity;

  for (let n of nums) {
    if ([first_max, second_max, third_max].includes(n)) continue;

    if (n > first_max) {
      third_max = second_max;
      second_max = first_max;
      first_max = n;
    } else if (n > second_max) {
      third_max = second_max;
      second_max = n;
    } else if (n > third_max) {
      third_max = n;
    }
  }

  if (third_max === -Infinity) return first_max;
  return third_max;
};

// Time: O(n)
// Space: O(1)
